### PR TITLE
Continued dielectric analysis module

### PR DESCRIPTION
**Pull request by PicoCentauri**
This PR continues the work of [PR#735](https://github.com/MDAnalysis/mdanalysis/pull/735). 

Adds the ability to calculate the static dielectric constant of a system. Furthermore also the directional dependent static constant can be calculated (useful if i.e. an electric field is applied). The code was compared to the gromaces `gmx dipole` module, giving the same results.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: testsuite/MDAnalysisTests/utils/test_units.py
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/utils/test_units.py]: ok this is confusing.  I tried changing this to the Å my keyboard has and it can't commit as it doesn't see any changes...

Comment (PicoCentauri) [reply testsuite/MDAnalysisTests/utils/test_units.py]: I just tried to reset the old utf character. Maybe it works....

**Comment by richardjgowers** [comment]
@PicoCentauri just an FYI, once this is "ready" it will have to wait to get merged until we've got a 0.19.1 release out.  As this is a new feature, it has to go into a 0.20.0 release

File: package/MDAnalysis/analysis/dielectric.py
Comment (jbarnoud) [inline package/MDAnalysis/analysis/dielectric.py]: I would like to see the units each value is in.

**Comment by jbarnoud** [comment]
It looks like something went wrong with your merge/rebase.

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: done

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: done

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: good idea! I added an option for making the wholes whole.

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: I never worked with duecredrit but I added this according to your suggestions. seems to look fine.

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: done

**Comment by PicoCentauri** [comment]
Sorry for the screwed rebase. It should be fine now.

File: package/MDAnalysis/analysis/dielectric.py
Comment (jbarnoud) [inline package/MDAnalysis/analysis/dielectric.py]: The test cannot the `False`. Here, `make_whole` refers to `MDAnalysis.lib.mdamath.make_whole` and a function always test as `True`. It should be `self.make_whole`, which should be defined in the `__init__`.

The fact that id did not make the tests fail indicate that it is not covered by the tests.

File: package/MDAnalysis/analysis/dielectric.py
Comment (jbarnoud) [inline package/MDAnalysis/analysis/dielectric.py]: The following lines (until the parenthesis that closes `due.cite`) should be aligned with `Doi` in agreement to PEP8.

File: package/MDAnalysis/analysis/dielectric.py
Comment (jbarnoud) [inline package/MDAnalysis/analysis/dielectric.py]: If there is no charge defined in the topology, this will likely raise a not so useful `AttributeError`. It would be nice to catch it, and raise a `NoDataError` with a friendly message, instead.

File: package/MDAnalysis/analysis/dielectric.py
Comment (jbarnoud) [inline package/MDAnalysis/analysis/dielectric.py]: You need to also import `absolute_import` from `__future__` to make pylint happy.

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: I fixed this and added a test. But I didn't found a nice test file where this actually makes a different for the results.

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: done

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: Good idea. I added this and also a check if the total charge is non zero

File: package/MDAnalysis/analysis/dielectric.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/dielectric.py]: done

**Comment by jbarnoud** [comment]
The test file has the molecule whole already. For the test to be useful, 
you need to break the molecules trhough PBC. You can try to wrap the 
molecules in the box first and see if they are broken. If they are not, 
you will have to translate them and wrap them in the box so that they break.

If you use the memory reader, you can do these operations in MDAnalysis 
at the beginning of the test.

**Comment by zemanj** [comment]
There is a fundamental problem in the code: It only checks if the entire selections total charge is zero.
This check is certainly required but not sufficient. If the system has PBC and contains molecules carrying a net charge, the analysis will yield wrong results.

* In case of a system containing net-neutral molecules only, the system's itinerant dipole moment can be expressed as the sum of molecular dipole moments. That's correctly dealt with by making molecules whole in every time step.
* However, in cases where ionic species are present, there exists a generally non-vanishing cumulatve *inter*molecular dipole moment, which unfortunately depends on the choice of the origin and is therefore undefined.
In such a cases, one has to calculate the relative dielectric permittivity from the itinerant dipole moment's temporal derivative, a.k.a. current (sum over charge times velocity instead of over charge times position).

**tl;dr**: It has to be checked that *every fragment* in the selection is charge neutral.
If not, one can either try to compute the dielectric permittivity using velocities (if present) or just throw an error with an appropriate error message.

**Comment by PicoCentauri** [comment]
@jbarnoud I was also unhappy with the test. Thanks for the hint. I will break some molecules and adjust the test.

@zemanj Thanks for pointing this out. I forgot to add this check. If you agree I will now only add a check for each fragment. But of course we can implement the term arising from current later.

**Comment by zemanj** [comment]
@PicoCentauri Sure, implementing routines for calculating the permittivity of systems with ionic species can be done later. Just be aware of the fact that checking all `selection.fragments` for charge neutrality won't suffice since the selection might include the fragments only partially.

In that respect, it would be cool to have an optional `compound` parameter for `*Group.total_charge()` enabling it to compute the total charge per segment/residue/molecule/fragment in the spirit of `*Group.center()`, i.e., taking into account only atoms *belonging to the group*.
Then, you'd simply have to check for `np.allclose(selection.total_charge(compound='fragments'), 0.0)`.

**Comment by PicoCentauri** [comment]
> it would be cool to have an optional compound parameter for *Group.total_charge()

Yes this seems very handy. I will look at your recent [commit](https://github.com/MDAnalysis/mdanalysis/commit/a2db387843874b669a63baa56cbb125fc6274486) and see what I can do.

File: testsuite/MDAnalysisTests/analysis/test_dielectric.py
Comment (PicoCentauri) [inline testsuite/MDAnalysisTests/analysis/test_dielectric.py]: This test should work with and without making the fragments whole. But it seems that the method `group.unwrap()` for making fragments whole does not work here - at least when applied once. By running
```
        eps_whole = DielectricConstant(ag, make_whole=True)
        eps_whole.run()
        eps_whole.run()
```
you get a result close to the desired one. Does somebody have an idea?

Comment (zemanj) [reply testsuite/MDAnalysisTests/analysis/test_dielectric.py]: @PicoCentauri You set up the `results` dict in the class initializer instead of the `_prepare()` method.
Thus, if you execute the `run()` method a second time, the `results` will already be filled with data from the previous run.
Initialize or reset the `results` dict in the `_prepare()` method and you should be fine.

File: package/MDAnalysis/units.py
Comment (richardjgowers) [inline package/MDAnalysis/units.py]: Is there a reason we can’t just import from SciPy.constants now? Less code to maintain. In fact I’d sooner just import from scipy in the actual module rather than us maintain our own list of constants

File: testsuite/MDAnalysisTests/analysis/test_dielectric.py
Comment (richardjgowers) [reply testsuite/MDAnalysisTests/analysis/test_dielectric.py]: These two expected results should be different test cases

File: package/MDAnalysis/units.py
Comment (jbarnoud) [reply package/MDAnalysis/units.py]: Until very recently, scipy was an optional dependency. I guess it would be fine now.

Comment (PicoCentauri) [reply package/MDAnalysis/units.py]: I’m also preferring using the scipy constants directly. The question for this PR is: Should we keep it like this for now and update it later?
Should we introduce scipy.constants for the new constants here? Or first adjust the units.py?

Comment (jbarnoud) [reply package/MDAnalysis/units.py]: I would say that changing how the constants are handled is out of scope for this PR.

**Comment by richardjgowers** [comment]
Sure it’d be nice to not further the (IMO) (post scipy being reqd) bad idea

File: testsuite/MDAnalysisTests/analysis/test_dielectric.py
Comment (PicoCentauri) [reply testsuite/MDAnalysisTests/analysis/test_dielectric.py]: I split the test case and also improved the creation of the results dictionary.

**Comment by PicoCentauri** [comment]
Do we still want to merge this PR? I'm open for changes or adjustments and it would be a bit sad throw our work away...

**Review by jbarnoud — APPROVED** [review]

**Comment by jbarnoud** [comment]
@PicoCentauri @richardjgowers @zemanj This PR looks like it has been sitting there for no particular reason. Is there anything blocking it going through?

File: package/MDAnalysis/analysis/dielectric.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/dielectric.py]:
```suggestion
:Year: 2022
```

File: package/MDAnalysis/analysis/dielectric.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/dielectric.py]:
```suggestion
    .. versionadded:: 2.1.0
```

**Comment by PicoCentauri** [comment]
No, I think it is ready.

**Review by IAlibay — CHANGES_REQUESTED** [review]
couple small things

**Comment by jbarnoud** [comment]
@IAlibay this needs you approval to go through.

**Comment by IAlibay** [comment]
So quick pre-approval question here - once we merge this code in, who holds the copyrights to this code? I.e. if we do go ahead with a re-license in the near future, do we have a means of contacting @mattiafelice-palermo ?

File: testsuite/MDAnalysisTests/analysis/test_dielectric.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_dielectric.py:5]: Should the copyright notice span the correct range of years? Not sure what our policy on this has been.

Comment (PicoCentauri) [reply testsuite/MDAnalysisTests/analysis/test_dielectric.py:5]: I also don't know. But at least we could put 2022.

Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_dielectric.py:5]: We've been super inconsistent here 😕 I guess it doesn't matter so much?

We mostly need a decision on https://github.com/MDAnalysis/mdanalysis/pull/2118#issuecomment-1032863726

Comment (mattiafelice-palermo) [reply testsuite/MDAnalysisTests/analysis/test_dielectric.py:5]: Dear @IAlibay, I apologize for being unresponsive but I have not been active in the field of computational chemistry for several years. Let me know what kind of approval you need in order to proceed with the PR.

Comment (jbarnoud) [reply testsuite/MDAnalysisTests/analysis/test_dielectric.py:5]: Hi @mattiafelice-palermo !

We are currently exploring if it is desirable and feasible to change the license of MDAnalysis to something more permissive. We will soon poll the contributors, but we need a way to reach them all.

We need a way to reach you for that poll or need to know if you are OK with your contribution to MDAnalysis to be relicensed to a more permissive open source license. If you are not OK with that, it is not a blocker, we'll just add that on the things to consider.

Cheers.

Comment (mattiafelice-palermo) [reply testsuite/MDAnalysisTests/analysis/test_dielectric.py:5]: @jbarnoud Thanks for the thorough explanation! You can reach me through GitHub, or if an email address is needed, I can PM it to you. I am more than ok with relicensing the code to a more permissive license :)

**Pull request merged**
